interp/iostream.go

```diff
@@ -0,0 +1,203 @@
+package interp
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"syscall"
+)
+
+const (
+	notClosedExitCode = -127
+)
+
+func cmdExitCode(cmdError error, ioerrs... error) (exitCode int, err error) {
+	var ee *exec.ExitError
+	var ok bool
+
+	if cmdError == nil {
+		// exitCode = 0.0
+		goto CheckIOErrors
+	}
+	if ee, ok = cmdError.(*exec.ExitError); !ok {
+		exitCode = -1.0
+		return
+	}
+	if exitCode, err = exitErrorCode(ee); err != nil {
+		return
+	}
+
+CheckIOErrors:
+	for _, err = range ioerrs {
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+// Copy the behavior of [gm]awk for pipe exit status. n?awk returns the raw
+// wait(3) return value but this is hard to use without a bit shift operator!
+func exitErrorCode(ee *exec.ExitError) (exitCode int, err error) {
+	stat, ok := ee.ProcessState.Sys().(syscall.WaitStatus)
+	if !ok {
+		// Not all platforms support WaitStatus?
+		return -1.0, err
+	}
+	switch {
+	case stat.Exited():
+		exitCode = stat.ExitStatus()
+	case stat.Signaled():
+		exitCode = 256 + int(stat.Signal())
+		if stat.CoreDump() {
+			exitCode += 256
+		}
+	default:
+		exitCode = -1
+		err = ee
+	}
+	return
+}
+
+type inputStream interface {
+	io.ReadCloser
+	ExitCode() int
+}
+
+type outputStream interface {
+	io.WriteCloser
+	Flush() error
+	ExitCode() int
+}
+
+type outFileStream struct {
+	*bufio.Writer
+	io.Closer
+	exitCode int
+}
+
+func outputFileStream(wc io.WriteCloser, size int) outputStream {
+	b := bufio.NewWriterSize(wc, size)
+	return &outFileStream{b, wc, notClosedExitCode}
+}
+
+func (s *outFileStream) Close() (err error) {
+	err1 := s.Writer.Flush()
+	err2 := s.Closer.Close()
+	switch {
+	case err1 != nil:
+		s.exitCode, err = -1, err1
+	case err2 != nil:
+		s.exitCode, err = -1, err2
+	default:
+		s.exitCode = 0
+	}
+	return
+}
```

Comment:
I think this is mostly just a personal aesthetic choice and I don't have a strong opinion about those things. I do see it as a choice between errors, after ensuring that `Flush()` and `Close()` are always run. To me, the suggestion is similar to the `if err := Etc(); err != nil { ... }` pattern which is more appropriate for sequential operations.

I will change it but I prefer the original ofc 😄 .

In all seriousness and not trying to be sarcastic, but if you use the named returns more often you might get used to them and not find them confusing. I am still getting used to them myself. The default values are a little odd sometimes... in other contexts as well... but I'm starting to like them.

Re: your comment on `cmdExitCode` ... right it isn't used here because the extra complexity of `*exec.Cmd` isn't an issue. The reason that `ioerrs` is an `[]error` slice is so that it can be used for consistent error handling for three different callers:

1. `outCmdStream.Close()` has 2 possible prior I/O errors
2. `inCmdStream.Close()` has 1 possible prior I/O error
3. `BuiltinSystem` (from vm.go) has 0 possible prior I/O errors

The slice makes handling all three different situations much simpler.

Reply:
I'm okay with the switch if you want to leave that, but let's 1) use `flushErr` and `closeErr` instead of `err1` and `err2`, and 2) avoid the naked returns and just do `s.exitCode = -1 \ return flushErr` type of thing.

Reply:
I've removed the named returns and there is now a check for double-closing in every Close() method:
```go
func (s *outFileStream) Close() error {
	if s.exitCode != notClosedExitCode {
		return doubleCloseError
	}
	flushErr := s.Writer.Flush()
	closeErr := s.closer.Close()
	if err := firstError(flushErr, closeErr); err != nil {
		s.exitCode = -1.0
		return err
	}
	s.exitCode = 0.0
	return nil
}
```